Composition with pretty pictures, more reorganization. (#1256)
square/workflow: docs/userguide/core-workflow.md

The overall state of a Workflow has two parts:

* [Props](../../glossary#props), configuration information provided by whatever is running the Workflow
* And the private [State](../../glossary#state) managed by the Workflow itself

At any time, a Workflow can be asked to transform its current Props and State into a [Rendering](../../glossary#rendering) that is suitable for external consumption.
A Rendering is typically a simple struct with display data, and event handler functions that can enqueue [Workflow Actions](../../glossary#action) — functions that update State, and which may at the same time emit [Output](../../glossary#output) events.

![Workflow schematic showing State as a box with Props entering from the top, Output exiting from the top, and Rendering exiting from the left side, with events returning to the workflow via Rendering](../images/workflow_schematic.svg)

A workflow Rendering usually serves as a view model in iOS or Android apps, although that is not a requirement.
(In fact, this page includes no details about how platform specific UI code is driven.
See [Workflow UI](../ui-concepts) for that discussion.

> And readers with an Android background should note the lower case _v_ and _m_ of "view model" — this notion has nothing to do with JetPack `ViewModel`.)

Author: Nice, thanks.

> And readers with an Android background should note the lower case _v_ and _m_ of "view model" — this notion has nothing to do with JetPack `ViewModel`.)

Author: Also added an extra paragraph to provide more context for that `GameWorkflow` pic:

```
For example, a Workflow running a simple game might be configured with a description of the participating `Players` as its Props, build `GameScreen` structs when asked to render, and emit a `GameOver` event to signal that is finished.

![Workflow schematic with State type GameState, Props type Players, Rendering type GameScreen, and Output type GameOver. The workflow is receiving an onClick() event.](../images/game_workflow_schematic.svg)

A workflow Rendering usually serves as a view model in iOS or Android apps, but that is not a requirement.
In fact, this page includes no details about how platform specific UI code is driven.
See [Workflow UI](../ui-concepts) for that discussion.

!!! note "Names are hard"
     Readers with an Android background should note the lower case _v_ and _m_ of "view model" — this notion has nothing to do with Jetpack `ViewModel`.
```

> And readers with an Android background should note the lower case _v_ and _m_ of "view model" — this notion has nothing to do with JetPack `ViewModel`.)

Reviewer: nice!

> ![Workflow schematic with State type GameState, Props type Players, Rendering type GameScreen, and Output type GameOver. The workflow is receiving an onClick() event.](../images/game_workflow_schematic.svg)

Reviewer: I'm not sure what to make of this image's placement.  It seems like it's meant to be a real-world example of the abstraction in the previous diagram, but since they're split by a paragraph, the connection is a little hard to make.

Reviewer: Actually with the diagrams and supporting text in the next section, I'm not sure this image is necessary at all?

Author: WDYT with the new paragraph above? Does it work now, or should I delete it and the image?

Reviewer: I like it.  It's nice baby steps.

* Has no private State, and emits no Output

![Workflow schematic showing MessageWorkflow](../images/email_message_workflow_schematic.svg)

**EmailBrowserWorkflow**

* State includes a `List<MessageId>`, and the selected `MessageId`
* Rendering is a `SplitScreen` view model, to be assembled from the renderings of the other two Workflows
* Accepts no Props, and emits no Output

![Workflow schematic showing MessageWorkflow](../images/email_browser_workflow_schematic.svg)

When `EmailBrowserWorkflow` is asked to provide its Rendering, it in turn asks for Renderings from its two children.

* It provides the `List<MessageId>` from its state as the Props for `EmailInboxWorkflow` and receives an `InBoxScreen` rendering in return. That `InboxScreen` becomes the left pane of a `SplitScreen` Rendering.
* For the `SplitScreen`'s right pane, the browser Workflow provides the currently selected `MessageId` as input to `EmailMessageWorkflow`, to get a `MessageScreen` rendering.

![Workflow schematic showing EmailBrowserWorkflow rendering by delegating to two children, InboxWorkflow and MessageWorkflow, and assembling their renderings into its own.](../images/split_screen_schematic.svg)

> (Note that the two children, `EmailInboxWorkflow` and `EmailMessageWorkflow`, have no knowledge of each other, nor of the context in which they are run.)

Author: Yeah, I like that.

![Workflow schematic showing State as a box with Props entering from the top, Output exiting from the top, and Rendering exiting from the left side, with events returning to the workflow via Rendering](../images/workflow_schematic.svg)

For example, a Workflow running a simple game might be configured with a description of the participating `Players` as its Props, build `GameScreen` structs when asked to render, and emit a `GameOver` event to signal that is finished.

![Workflow schematic with State type GameState, Props type Players, Rendering type GameScreen, and Output type GameOver. The workflow is receiving an onClick() event.](../images/game_workflow_schematic.svg)

A workflow Rendering usually serves as a view model in iOS or Android apps, but that is not a requirement.
In fact, this page includes no details about how platform specific UI code is driven.
See [Workflow UI](../ui-concepts) for that discussion.

!!! note
     Readers with an Android background should note the lower case _v_ and _m_ of "view model" — this notion has nothing to do with Jetpack `ViewModel`.

## Composing Workflows

> Workflows run in a tree, with a single root Workflow spawning any number of children, each of which can spawn children of their own, and so on.

Reviewer 2: ```suggestion
Workflows run in a tree, with a single root Workflow declaring it has any number of children for a particular state, each of which can declare children of their own, and so on.
```

* Accepts no Props, and emits no Output

![Workflow schematic showing MessageWorkflow](../images/email_browser_workflow_schematic.svg)

When `EmailBrowserWorkflow` is asked to provide its Rendering, it in turn asks for Renderings from its two children.

* It provides the `List<MessageId>` from its state as the Props for `EmailInboxWorkflow` and receives an `InBoxScreen` rendering in return. That `InboxScreen` becomes the left pane of a `SplitScreen` Rendering.
* For the `SplitScreen`'s right pane, the browser Workflow provides the currently selected `MessageId` as input to `EmailMessageWorkflow`, to get a `MessageScreen` rendering.

![Workflow schematic showing EmailBrowserWorkflow rendering by delegating to two children, InboxWorkflow and MessageWorkflow, and assembling their renderings into its own.](../images/split_screen_schematic.svg)

!!! note
    Note that the two children, `EmailInboxWorkflow` and `EmailMessageWorkflow`, have no knowledge of each other, nor of the context in which they are run.

The `InboxScreen` rendering includes an `onMessageSelected(MessageId)` function.

> When that is called, `EmailInboxWorkflow` emits the given `MessageId` as Output to be processed by its parent.

Reviewer 2: More accurate/clear with definitions established above but possibly too complex for here?
```suggestion
When that is called, `EmailInboxWorkflow` sends an action to the runtime that emits the given `MessageId` as Output to be processed by its parent.
```

> `EmailBrowserWorkflow`  receives that output, and updates the `selection: MessageId` of its State accordingly.

Reviewer 2: ```suggestion
`EmailBrowserWorkflow`  receives that output, and sends an action to the runtime that updates the `selection: MessageId` of its State accordingly.
```

Author: > The `InboxScreen` rendering includes an `onMessageSelected(MessageId)` function.
When that is called, `EmailInboxWorkflow` enqueues an Action function that emits the given `MessageId` as Output.
> `EmailBrowserWorkflow`  receives that output, and enqueues another Action that updates the `selection: MessageId` of its State accordingly.

How's that?

![Workflow schematic showing EmailBrowserWorkflow rendering by delegating to two children, InboxWorkflow and MessageWorkflow, and assembling their renderings into its own.](../images/split_screen_update.svg)

> Whenever such a [Workflow Action cascade](../../glossary#action-cascade) fires, the root Workflow is asked for a new Rendering.

Reviewer 2: Since you use 'Action Cascade' explicitly here - I think we should specify actions above, otherwise the gap is too big.

Author: Yeah, I struggled with that. This is an improvement, thanks.
